### ISSUE TITLE
Discussion on the transposing trick used in transformer

### ISSUE THREAD

**Issue by Joy-Lunkad** [openai/CLIP#280]
I was hoping to understand the benefits of the "double transpose trick" employed in [model.py](https://github.com/openai/CLIP/blob/main/clip/model.py) and its potential effects on layernorm inside and outside the transformer.

I would greatly appreciate if you could point me towards a good resource that explains this.

```python3
x = self.ln_pre(x)
x = x.permute(1, 0, 2)  # NLD -> LND
x = self.transformer(x)
x = x.permute(1, 0, 2)  # LND -> NLD
x = self.ln_post(x[:, 0, :])
```

Thank you!

**Comment by jongwook** [comment]
Hi, it was just to conform with the input shape expected by the `nn.MultiheadAttention` module, which is `(L=sequence length, N=batch size, D=model width)` by default. I just noticed that PyTorch has added the `batch_first=True` option in 1.9.0 which would eliminate the need to transpose.

**Comment by Joy-Lunkad** [comment]
Thank you for your reply! I have just one more question :)
LN inside transformer:-

```python3
self.ln_1 = LayerNorm(d_model)
```
LN outside transformer:-
```python3
self.ln_pre = LayerNorm(width)
```
Am I missing something obvious here since these do not inform the [layer norm module](https://pytorch.org/docs/stable/generated/torch.nn.LayerNorm.html
) which axis is the batch axis that it needs to ignore?

**Comment by jongwook** [comment]
The LayerNorm module will use the last dimension(s), and in our case the model width is always the last dimension before or after the transposes.

It was pointed out in #152 (with a similar discussion about the redundancy of `ln_pre`) that common implementations of vision transformers don't use `ln_pre`. Model-wise, there's a slight difference with and without `ln_pre`, but I think it wouldn't make much difference when trained without `ln_pre`.
